README.md

```diff
@@ -13,23 +13,27 @@
 
 *Plutonium Addon: Automation* runs invisibly and mostly silently.
 
-While the module is activated in your world, each time Plutonium imports something, it'll check to see if any automation data is set for that thing. If there is, it will import the item with that data. If not, it will import using the Plutonium standard, filling out fields as a reasonable guess.
+While the module is activated in your world, each time Plutonium would import a document, it will check if any automation data is available for that document. If so, it will import the document with the additional automation data included. If not, it will import the document as Plutonium standard (although sometimes slightly tweaked to better enable the module dependencies outlined below).
 
 As with Plutonium, **all data is included in the module itself**. There's no need to import everything prematurely, 'just in case'.
 
-Some effects don't use effects directly, but rather call hidden macros. To view/edit these macros, install and activate [Item Macro](https://github.com/Kekilla0/Item-Macro).
+Some effects don't use effects directly, but rather call item macros. If you, as the GM, wish to allow players to view/edit these macros on items they own, configure Item Macro to allow the _Player access_ option.
 
 ### Dependencies and incompatibilities
 
 Besides Plutonium itself, *Plutonium Addon: Automation* **requires**:
-- [DAE](https://gitlab.com/tposney/dae)
-- [Midi QoL](https://gitlab.com/tposney/midi-qol)
-- [times-up](https://gitlab.com/tposney/times-up)
-- [DFred's Convenient Effects](https://github.com/DFreds/dfreds-convenient-effects)
+- [DAE](https://foundryvtt.com/packages/dae)
+- [Midi QoL](https://foundryvtt.com/packages/midi-qol)
+- [times-up](https://foundryvtt.com/packages/times-up)
+- [DFred's Convenient Effects](https://foundryvtt.com/packages/dfreds-convenient-effects)
+- [Item Macro](https://foundryvtt.com/packages/itemacro)
 
-Some specific items require other modules. You will only be prompted to install/activate these modules if you import an item that requires it.
+Some specific automations make use of additional modules, although these are not required. These modules are:
```

Comment:
Actually, one thing regarding this: is this _not_ going to be prompted? I understand that it's not coded right now but without the "THIS DOCUMENT REQUIRES [THIS] MODULE—INSTALL/ENABLE IT NOW" popup I think it's hard to really call them 'not required'.

Reply:
well it's currently just a lie so I thought I'd change it 😏 

a fine point to discuss, though

- since we "can't" (ignoring Plutonium backend modding) install modules mid-game, prompting the user to take action, but not giving them a direct action to take, is a bit ugly. Could pop up a notification per-module, once per session? I'm envisioning the bad case of "I import thing which requires X. I dismiss the alert as there is no other action available, and I do not want to return to setup right now. I import another thing which requires X+Y. I dismiss the alert... ..."
- If we are prompting/requiring a user to install something, would it not be better just to make it a dependency..?
- If we are happy for a user to miss out on \<x functionality\> because it is optional, then why are we prompting them to install things?
- do we have multiple versions of something, and degrade to the version that works for the user? This is probably the "best" solution, but it's much more complex, and a total pain (and may not produce meaningful improvements in ever case). (The burden of creating/maintaining automations is already very high, so I don't see much of a future in adding to it where it can be avoided)
- Instead of prompting, could we change the `requires: {"module": true}` to `requires: {"module": "reason"}` and generate a docs page (in-game, out-of-game, both), with some basic search/type filtering so a user could check and see what they might be missing out on if they don't install X module?
   - Or, even simpler, skip the reasons and just generate links to the optional modules and stick them in a dialogue accessible via the PAA module settings in-game. That way, if a GM is going through their configs tweaking things, there's a fair chance they'll go "hm what's this" and click the "View Optional Dependencies" button, and read/install from there

My personal expectation was that we'd do something in a similar style to e.g. the Warp Gate+Sequencer+JB2A+etc. else stack, whereby each module doesn't _actually_ depend on the others, but if you have them all, suddenly your spells get cool animations. i.e. it's all silent, but magically Just Works:tm: if you're running the right stuff anyway

Reply:
> I'm envisioning the bad case of "I import thing which requires X. I dismiss the alert as there is no other action available, and I do not want to return to setup right now. I import another thing which requires X+Y. I dismiss the alert... ..."

This is true, but in my bluntly honest opinion, that's a user error. Not saying a user should return to the setup screen *immediately* upon importing something, but to not make at least a mental note to check later sounds silly. An alternative (or concurrent!) solution would be to post a _chat message_ (which are slightly more persistent) with a reminder to enable the modules.

Example workflow:
- When a document is imported:
  - If the corresponding PAA data contains a `requires` array, write to a world flag\* the list of dependencies required, discarding duplicates.
  - Create a warning toast if a required module is missing.
- When the importer is closed, and when the world is loaded:
  - Test the aforementioned flag against currently enabled modules. If a module is missing, write a private chat message between the current user and the GM informing them to install/enable the missing module(s). (Ideally have some way to auto-delete older outdated/redundant messages to avoid spam.)

  \* idk how the data is actually stored but, whether it's in an invisible PAA setting or an actual flag, basically just store it _somewhere_

---

> If we are prompting/requiring a user to install something, would it not be better just to make it a dependency..?

An easy example would be Active Auras. There are only a very small handful of 'aura effects' in official 5e material—the paladin class and a few aptly named spells are the only ones I know of—so you can very plausibly run an entire years-long campaign never needing this module (my own group didn't need it for the first 2 years of weekly play!). Generally, enabling redundant modules is ill-advised:
1) Active Auras is quite computationally intensive. Even if you don't have any aura AEs, I believe it's still being called for every single token-movement (for any token).
2) Setting spam: I may be obsessive enough to go through every single configuration button of every single module, but most won't. If you don't check Active Aura's settings, for instance, then you'll find it gets called for every token-movement... even outside combat.
3) More modules required = more modules that we need to check are compatible at every stage. If `dnd5e` or Foundry update, PAA can't reasonably update until Active Auras does too, or we may be forcing users to be running outdated/buggy code they never even needed.

---

> do we have multiple versions of something, and degrade to the version that works for the user?

I say hard :No: to that, just for sanity alone 🤣

If someone else wants to maintain alternatives, that's their own prerogative. But, especially in the case of macros, I do not want to be doubling or even tripling the work.

---

> [last bullet and sub-bullet]

This is a good idea. If you don't like big warnings, you can amend my above workflow to instead just post a warning toast on importer-close, scene-change, and world-load (i.e. plentiful reminders) that says "Some modules are missing. See PAA settings for more information." That would be much 'cleaner', and would

Reply:
I'm gonna respond in a slightly different order to your bullets so I sound coherent. Bear with me as I try to be thorough pls; doing high-level discussion first and then implementation later 😩

---

> do we have multiple versions of something, and degrade to the version that works for the user?

I say hard :No: to that, just for sanity alone 🤣

If someone else wants to maintain alternatives, that's their own prerogative. But, especially in the case of macros, I do not want to be doubling or even tripling the work, and I'm near-certain that everyone else contributing to this repo would mirror the sentiment.

---

> If we are prompting/requiring a user to install something, would it not be better just to make it a dependency..?

To illustrate my thinking, I'll use the example of Active Auras. There are only a small handful of 'aura effects' in official 5e material—the paladin class after 6th level and a few aptly named spells are the only ones I know of—so you can very plausibly run a years-long campaigns never needing this module (my own group didn't need it for ~2 years of weekly play!). Enabling a redundant module is imo fairly ill-advised:
1) Active Auras is quite computationally intensive. Even if you don't have any aura AEs, I believe it's still being called for every single token-movement (for any token). Not gonna brick your browser, but how many redundant CPU cycles will be forced by PAA by hard-requiring _all_ helper modules?
2) Setting spam: I may be obsessive enough to go through the entire page for every single module, but most won't. If you don't check Active Aura's settings, for instance, then you'll find the above occurs for every token-movement... even outside combat.
3) More modules required = more modules that we need to check are compatible at every stage. For instance, Active Auras took a few months to migrate to v10 iirc, and Kandashi has been fairly inactive as of late. Forcing a dependency would make him a 'single point of failure' for all users, rather than just those actually using aura effects. (No disrespect at all to Kandashi if you happen to be reading this btw, just using an example 🥰)

Yes, we _can_ just list all the modules as hard dependencies, and it _would_ be easier on our end. However, for each helper module we end up using _anywhere_ in PAA data, it gts inflicted on every user indiscriminately.

The current dependencies are only there because I don't think it's possible/feasible to play even a single session without once needing to apply conditions (DFCE) or start-/end-of-turn effects (Times Up); 5e just has too many of them. Auras, summons, etc. are quite empirically not the case.

---

> My personal expectation was that we'd do something in a similar style to e.g. the Warp Gate+Sequencer+JB2A+etc. else stack, whereby each module doesn't actually depend on the others, but if you have them all, suddenly your spells get cool animations.

The thing is I'm not convinced that really works here. If you have Sequencer and JB2A, you can place animations on the screen manually, but without Automated Animations it doesn't trigger automatically. You can have a fancy _dancing lights_ animation occur as expected with all three, but without Warp Gate you can't spawn in a temporary actor to represent the lights. If you only have Warp Gate, you _can_ spawn in the actors, but they'll just instantly appear out of the ether rather than follow a VFX 'script'.*

The point is that each module in that set does something fundamentally separate. However, PAA is trying to automate mechanics holistically: if you have a paladin but don't enable Active Auras... then it just doesn't work. Midi-QoL will apply an AE that references Active Auras, but Active Auras won't be around to handle it, so the AE will either just not work or start spamming errors all over.

The only way to avoid this, without maintaining multiple PAA data sets as above, is to only import PAA data _if_ Active Auras is enabled... which imo is even worse UX than importing the automation available and then reminding the user to enable the helper module.

\* Most people don't actually do this; they rely on big 'autorec' files filled with macros for all sorts of documents. These autorec files _already assume_ the other modules as a dependency—they just don't have a way to enforce it like modules do.

So, re:
> If we are happy for a user to miss out on <x functionality> because it is optional, then why are we prompting them to install things?

I'm not happy for the user to miss out on the functionality, because that would be identical to just importing the Plutonium standard!

---

> I'm envisioning the bad case of "I import thing which requires X. I dismiss the alert as there is no other action available, and I do not want to return to setup right now. I import another thing which requires X+Y. I dismiss the alert..."

This is true, but, in my bluntly honest opinion, that's a user error. Not saying a user should return to the setup screen *immediately* upon importing a document, but to not make at least a mental note to check later sounds like an oversight. An alternative (or concurrent!) solution would be to post a _chat message_ (which are slightly more persistent) with a reminder to enable the modules.

Example workflow:
- When a document is imported:
  - If the corresponding PAA data contains a `requires` array, write to a world flag\* the list of dependencies required, discarding duplicates.
  - Create a warning toast if a required module is missing.
- When the importer is closed:
  - Post warnings in a private chat message between the current user and the GM. (Ideally have some way to auto-delete outdated/redundant messages to avoid spam.)
- Add a button in PAA settings called "Recalculate Dependencies" or something, which iterates through each item and items in actors to build a new list of dependencies. This would be used if you end up deleting a bunch of stuff. Usual warning of "this might take a while" of course.

  \* idk how the data is actually stored but, whether it's in a PAA setting that says "DON'T EDIT UNLESS YOU KNOW WHAT YOU'RE DOING" or an actual flag, basically just store it somewhere

---

> generate a docs page (in-game, out-of-game, both), with some basic search/type filtering so a user could check and see what they might be missing out on if they don't install X module? 

This is a good idea in general tbh, and feeds into a suggestion (that the Discord user never added to Github 😤) about creating a big index of things that have _any_ automation. One can add a "requires [X]" line to appropriate entries there, and honestly even lever existing 5eT list pages' filter code to let it serve double-purpose.

I don't think the reasons are important; it should either be patently obvious (e.g. a paladin's Aura of Protection needs Active Auras) or something the end-user can figure out themselves by reading the PAA data. The people who would benefit from the reasoning would be coders, after all.

> Or, even simpler, skip the reasons and just generate links to the optional modules and stick them in a dialogue accessible via the PAA module settings in-game. 

If you don't like warning spam or writing messages, you can amend my above workflow to instead just post a single warning toast on importer-close and world-load (and scene-change?) that says "Some modules PAA requires are missing. See PAA settings for more information." That would be much 'cleaner' from a non-GM's perspective, but ultimately I think it's functionally the same. I kinda favour this, but ultimately I'll go with whichever is easiest to code/maintain; GMs should only need to really deal with these pop-ups once or twice in a campaign, anyway.

Reply:
If you prefer to discuss this on Discord for ease I'm up for that over writing essays 🤣

Reply:
![image](https://user-images.githubusercontent.com/25264887/218082518-76b538b6-0d24-4931-9c20-cf278d9f7bbe.png)

---

nah that all makes sense. 

- Agreed that "multiple datasets" is too painful to do
- Good point about the paladin/performance/general needless clutter
- Happy to have my "animation stack" idea shot down because the UX does kinda suck

As for the implementation details, blegh. Might have to go with "notification on import + chat message once-per-session-per-module," with config to disable notifications for each module, for now.

Your "world flag" idea is another version of the "there needs to be a way to re-import stuff to update it" idea, I think, which is a much bigger and nastier task (and more general to Plutonium)--ultimately that'll be some sort of diffing mechanism. If one has PAA running and diffs "existing import" vs "expected from fresh import" the missing automations would implicitly be flagged up, so that'll all come out in the wash eventually.

> creating a big index of things that have any automation

feel free to create the Issue, o essay writer 😏

Reply:
a rightfully deserved mucho texto 😩

---

> Your "world flag" idea is another version of the "there needs to be a way to re-import stuff to update it" idea

Is it? It _could_ be, but I don't think it _needs_ to be; it could be much, much simpler. I'll just spell it out in very-pseudo-JS/Foundry API to simplify words. The 'world flag' is here:
```js
// importer is opened
const requiredModules = PAA.flags.requiredModules;

// check-boxes are selected and the 'import' button has just been clicked
selectedDocuments.forEach(d => {
  // Plutonium.importer.doImportEtc();
  // PAA.importer.injectAutomationData();
  if (PAA.importedAutomationData.requires.count) requiredModules.push(PAA.importedAutomationData.requires);
});

// selected items are imported and has just finished
PAA.flags.write("requiredModules", requiredModules.flatten().unique());

// importer is closed or world is loaded
const activeModules = game.getEnabledModules();
const missingModules = [];
foreach (const module of PAA.flags.requiredModules) {
  if (!activeModules.find(module)) {
    missingModules.push(module);
    ui.notifications.warn(`pls enable ${module} thnx`);
  }
}
```
I don't think the above represents any significant computational effort compared to what Plutonium is already doing.

Here's something for the 'recalculate dependencies' button:
```js
const data = PAA.loadData();
const requiredModules = [];
foreach (const document of (items + actors.map(a => a.items).flatten())) {
  const automationData = data[document.dataType].find(i => i.name === document.name && i.source === document.source);
  if (automationData) requiredModules.push(automationData.requires);
}
requiredModules = requiredModules.flatten().unique();
const activeModules = game.getEnabledModules();
const missingModules = [];
foreach (const module of requiredModules) {
  if (!activeModules.find(module)) {
    missingModules.push(module);
    ui.notifications.warn(`pls enable ${module} thnx`);
  }
}
if (missingModules.count) chat.createPrivateMessage(`pls enable ${missingModules.join(" and ")} thnx`);
PAA.flags.write("requiredModules", requiredModules);
```

This might not be fast (I'd wager at order of minutes for a normal campaign world), but it would certainly do the job. Throw a warning next to the button that says "this might take a while" and count the number of documents to iterate through at the start to throw up a progress bar and suddenly the UX is fine. I know the above code is overly simplifying the API/data structure, but the only thing I'm _really_ uncertain about is the `document.dataType` (which might not be possible without a custom Plutonium flag); it all seems very doable without needing a diff.

Fully accept that I have little to no experience here though so if I'm just wrong then rip 😔

But even setting that aside, I think the 'update content in place' thing is already kinda sorted in theory? You did it for brew sources on 5eT: add a version number or last-modified timestamp, and do the same iteration as above, now comparing that number/timestamp against Plutonium's data. The rare case of overwriting content that was edited by a user after import can be dealt with by just putting a warning up, telling the user to rename either the name or source to something unique (e.g. "Fireball*" or "My Homebrew" respectively).

---

Also, re: this actual PR's topic
Item Macro now needs to be a non-hard dependency 🤣 

It's only used in one feature (Rage) so far, and 'the party doesn't have a barbarian' is sufficiently common (more than half the groups out there for sure) that I don't think Item Macro is really warranted as mandatory. I'll reverse if it ends up being used on 5+ or so core classes, if @nickelpro can foresee that far into the future for implementation.

Reply:
I think it's a very nice [bikeshed](https://en.wikipedia.org/wiki/Law_of_triviality) that we can paint any color we so desire, and if we decide to change the color in the future it's like a single find-and-replace.

Reply:
😏 not _quite_ bikeshedding, as what I'm really trying to uncover, here, is how I should implement the "optional dependency negotiator" system. Admittedly this PR is a questionable place to be writing out all the notes, but we're already 2 essays deep so why stop now.

---

@Spappz riiiight, so the idea is: import _everything_, then strongarm the user into installing and enabling modules, or the stuff they just imported won't work? I don't like that idea much 🤔 And most of my thoughts on how to get that idea to a point I _do_ like would involve the "multiple datasets" thing

Reply:
Importing something that relies upon something that is not installed, should be a definite no IMO. Especially for core features of classes. 
If it creates a regression like "I can't click the roll button on a feature" because some macro is trying to go off and erroring because a module isn't installed, then that's highly undesirable and is a great way to get people to not want the module at all. Some people may not want a specific module, and that's fine, but we shouldn't take a "well f you then, fix our broken import yourself".

Reply:
> always import every automation, then strongarm the user into installing and enabling modules, or the stuff they just imported won't work?

Yes. The reason to not list all helper modules as hard dependencies is to... reduce the number of dependencies, for the reasons I outlined above. A user who wants to use PAA to import automated documents should _always_ install and enable the modules their imported PAA documents require. There is zero use-case where they want the automation but not the module, because the module is the only thing that _allows_ the automation. If they just don't want automation on an item, then default, vanilla Plutonium is right there.

The only question, as I see it, is how the UX appears: do we politely request with the odd warning, spam messages until it's done, compile a dynamic list hidden in the settings, etc. It's (was) already in the (fairly short) README, so honestly RTFM does apply; a user who _notices_ the advisory that they need to install a module _but ignores it anyway_ is only shooting themselves in the foot.

If I'm misunderstanding a lot of implementation details and this really is too complex, just go ahead and make all the 'not quite so necessary' modules like Active Auras and Warpgate actual hard dependencies. ¯\\\_ (ツ)\_/¯

Reply:
I see. If we want to go with (1) "optional" as in _actually_ optional, then I think "notify now + allow recovery (by re-import) later," which is going to feed into a planned Plutonium system anyway, is the way to go.

If we want to go with (2) "optional" as in "optional until later required," then hard dependencies are going to be the best for both us and for the user, I think.

The latter is currently easier, but the former scales a bit better, and is ultimately going to be "free." Given that I already started on (1) in #78, and that adding dependencies to an existing Foundry module is afaik a mess, I'm putting my vote for (1) 😏

Reply:
Agree with what you said about (1), but I stand by what I said about maintaining two or more sets of data. I'd rather just go down the route of "if dependencies are missing, import the vanilla Plutonium default" and then create a reimport mechanism, rather than "if dependencies are missing, import a partial/stripped-down/alternative PAA version of the data that has to be maintained parallel to the main one".

I still vastly prefer the principle for (2), and agree that hard dependencies are easiest _right now_... but I'd still like to move to the _pragmatics_ of (1) because it's a better UX imo.

What's wrong with adding dependencies, out of interest? Does Foundry not properly check dependencies on world-load and disable a module if its dependencies are missing?

Reply:
> Does Foundry not properly check dependencies on world-load and disable a module if its dependencies are missing?

got it in one!